### ISSUE TITLE
Improve getting started experience for screen reader users

### ISSUE THREAD

**Issue by meganrogge** [microsoft/vscode#223787]
@rperez030 shared some feedback on the current getting started experience for screen reader users:

- Knowing which settings to configure is a task, ideally there'd be a wizard to set settings for them
- IntelliSense is overwhelming
- Easy to miss getting started experience, consider a dialog that provides options: open documentation, custom walkthrough, don't show again

There are some components worth calling out in an accessibility walkthrough: 
- Accessibility help
- Accessible view
- Settings
- IntelliSense
- Hover 

Our screen reader detection isn’t 100%, so concerned about showing a dialog, can we get some data?

Custom documentation with ways to action information:
https://github.com/microsoft/vscode/issues/223786
https://github.com/microsoft/vscode/issues/213417

also cc @isidorn and @jooyoungseo

**Comment by meganrogge** [comment]
discussed with JooYoung, who had similarly not tried our getting started experience due to a lack of discoverability.

We discussed the following categories:

code navigation, editing
- hover
- go to symbol
- accessible view
- code folding
- IntelliSense

discoverability
- verbosity settings
- accessibility help
- accessibility signals
- command palette

**Comment by meganrogge** [comment]
@jooyoungseo, can you help me with understanding the commands you find most useful for code folding and how they help you? I would imagine that you fold certain regions and then navigate them bw them using `Go to next/previous folding range`?

Re intellisense, I could also use your help. 

Also cc @rperez030 for guidance/wisdom here.

**Comment by meganrogge** [comment]
I have a getting started walkthrough underway and have noticed some issues.
- focus change - if a command is run from the walkthrough step, focus changes and a user will have difficulty getting back to the walkthrough/ particular step
- actions that are featured about aren't actionable from within the walkthrough - for example, trigger suggest, fold code, etc. due to their `when` clauses
- a user cannot configure a keybinding from within here for the commands that we mention

Idea:

Add the important editor commands to the editor's help dialog, where they are actionable and the keybindings are configurable within the accessibility help dialog:
- hover
- go to symbol
- code folding
- IntelliSense

Have the custom getting started walkthrough feature:
- command palette
- accessibility settings
- accessibility signals
- verbosity settings

**Comment by rperez030** [comment]
> I have a getting started walkthrough underway, including all of the above that we discussed, and have noticed some issues.
> 
> * focus change - if a command is run from the walkthrough step (for example, open accessibility settings), focus changes and a user will have difficulty getting back to the walkthrough/ particular step

i'm actually having the oposito problem. If I select Open settings, the settings UI grabs focus for a moment, then focus returns to my previous place in the walk-through before I have time to do anything. this is with the getting started walk-through.

> * actions that are featured aren't actionable from within the walkthrough - for example, trigger suggest, fold code, etc. due to their `when` clauses

I'm not sure that is a problem as long as we set expectations, and as long as the actions work as described in code.

> * a user cannot configure a keybinding from within here for the commands that we mention
> 
> Idea:
> 
> Add the important editor commands to the editor's help dialog, where they are actionable and the keybindings are configurable within the accessibility help dialog:

I think they can be in both places, even if they are not configurable from the walk-through. I believe changing key-bindings is something one is more likely to do after using the product for a period of time, not during the getting started experience.

> 
> * hover
> * go to symbol
> * code folding
> * IntelliSense
> 
> Have the custom getting started walkthrough feature:
> 
> * command palette
> * accessibility settings
> * accessibility signals
> * verbosity settings
> 
> Also, we may need an accessible view for these walkthrough steps because they're rendered markdown just like the chat responses.

The walk-through is presented to me as rendered HTML. this is perfect in this context becuse screen readers expose all the structural elements such as headings and lists, as well as the actionable elements. It is rendered inside an element with document role, so windows screen readers automatically virtualize that content. i don't think we need the accessible view.

There is something very problematic with the markup of the walk-through. as soon as I expand a collapsible item, everything inside appears as being expanded, and NVDA loses the ability to interact with the content. Without inspecting the code, I suspect the expanded content is added inside the button tag with the expanded attribute, which according to specification would make all the expanded content presentational / not actionable by assistive technology.

Here is an example of an accessible implementation of the accordion pattern: [https://www.w3.org/WAI/ARIA/apg/patterns/accordion/](https://www.w3.org/WAI/ARIA/apg/patterns/accordion/)

**Comment by meganrogge** [comment]
> i'm actually having the oposit problem. If I select Open settings, the settings UI grabs focus for a moment, then focus returns to my previous place in the walk-through before I have time to do anything. this is with the getting started walk-through. I would expect to be able to close settings by pressing CTRL +W, then return to the walk-through. This of course depends on the ability of the screen reader to maintain focus position in the previous tab.

Hmm, that is odd. When I focus the `Open Settings` button and hit enter, the settings editor opens to the side and the settings input is focused. When I use `cmd+w` (mac), focus returns to the walkthrough, but not the step (which is what @bhavyaus is planning to fix). Maybe there's a platform difference? I will try on windows.


> I'm not sure that is a problem as long as we set expectations, and as long as the actions work as described in code.

I still worry about including tabbable elements that seem like they should run an action and yet don't. Perhaps adding this to the description is sufficient as you suggested. Thanks for the feedback.

> > 
> > Idea:
> > Add the important editor commands to the editor's help dialog, where they are actionable and the keybindings are configurable within the accessibility help dialog:
> 
> I think they can be in both places, even if they are not configurable from the walk-through. I believe changing key-bindings is something one is more likely to do after using the product for a period of time, not during the getting started experience.

Once again, great feedback. Will add this to the editor's help dialog 👍 


> The walk-through is presented to me as rendered HTML. this is perfect in this context becuse screen readers expose all the structural elements such as headings and lists, as well as the actionable elements. It is rendered inside an element with document role, so windows screen readers automatically virtualize that content. i don't think we need the accessible view.

Okay good to know, cc @bhavyaus 
> 
> There is something very problematic with the markup of the walk-through. as soon as I expand a collapsible item, everything inside appears as being expanded, and NVDA loses the ability to interact with the content. Without inspecting the code, I suspect the expanded content is added inside the button tag with the expanded attribute, which according to specification would make all the expanded content presentational / not actionable by assistive technology.
> 
> Here is an example of an accessible implementation of the accordion pattern: https://www.w3.org/WAI/ARIA/apg/patterns/accordion/

**Comment by meganrogge** [comment]
Takeaways:
- add the content mentioned above to the editor's help dialog
- for cases when the button won't be actionable in the context, indicate that to the user for example "when an editor is focused"
- no need for an accessible view, at least per Roberto
- investigate this:
There is something very problematic with the markup of the walk-through. as soon as I expand a collapsible item, everything inside appears as being expanded, and NVDA loses the ability to interact with the content. Without inspecting the code, I suspect the expanded content is added inside the button tag with the expanded attribute, which according to specification would make all the expanded content presentational / not actionable by assistive technology.

Here is an example of an accessible implementation of the accordion pattern: https://www.w3.org/WAI/ARIA/apg/patterns/accordion/

**Comment by meganrogge** [comment]
@rperez030 thanks for mentioning this:

as soon as I expand a collapsible item, everything inside appears as being expanded, and NVDA loses the ability to interact with the content. 

can you provide the steps to reproduce the issue?

So you expand a step, then try to do what, which doesn't work?

**Comment by rperez030** [comment]
Here is how you can reproduce it.

1. Open VS code on Windows and make sure NVDA is running.
2. In the command palette, run "Welcome - Open Walkthrough.".
3. Press the letter B to navigate by buttons using the browse cursor. Observe that nVDA announces the buttons and corresponding checkboxes as a single control. (   check box  not checked    Sync settings across devices  button  collapsed).
4. Use letter X to navigate by checkboxes using the Browse cursor. Observe that NVDA just reads unlabeled checkboxes. (check box  not checked    button  collapsed ).
5. locate the button "Tune your settings", and press enter to expand the content.
6. Use the down arrow key to read the expanded content.

At this point, every single string of text in the expanded content is read as an expanded button, so it is not possible to tell what is actionable except by context. (button  expanded    Customize every aspect of VS Code and your extensions to your liking. Commonly used settings are). Also, it seems like it is not possible to collapse an element once expanded using the keyboard or the screen reader.

**Comment by meganrogge** [comment]
@jooyoungseo explained that using a tree view for this is odd from a screen reader user's perspective. Furthermore, he said that an accessible view would be needed. If users are accessing the walkthrough only to have to open the accessible view, we are better off using a custom editor.

**Comment by meganrogge** [comment]
Just tested and I now understand the issue. On macOS, with a walkthrough step focused, it reads the content of the step and then reads "You are on a button, group. To click this button, use `ctrl+option+space`". Doing so does nothing, which is confusing.

**Comment by meganrogge** [comment]
@jooyoungseo suggested adding an entry to our `Help` menu such as `Get Started with a Screen Reader`

**Comment by meganrogge** [comment]
@jooyoungseo, after more thought, I am not sure we want to diverge from the non-screen reader user's experience here. If possible, we should fix the existing one. 

Related to the accessible view, we can make sure that going between next and previous in the accessible view are supported similar to the other features.

**Comment by bhavyaus** [comment]
> Just tested and I now understand the issue. On macOS, with a walkthrough step focused, it reads the content of the step and then reads "You are on a button, group. To click this button, use `ctrl+option+space`". Doing so does nothing, which is confusing. [@bhavyaus](https://github.com/bhavyaus) could we use a more appropriate HTML tag here?

Please create a separate issue for this so that I can track it on my end.

**Comment by meganrogge** [comment]
That is tracked here https://github.com/microsoft/vscode/issues/225977
